File: src/sdk/cpp/ofa_agent/src/constraint.cpp
/**
 * @file constraint.cpp
 * @brief Constraint Checking Implementation
 * Sprint 30: Constraint Engine
 */

#include "ofa/constraint.hpp"
#include <chrono>
#include <algorithm>

namespace ofa {

static int64_t nowMs() {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()
    ).count();
}

// ==================== ConstraintChecker ====================

ConstraintChecker::ConstraintChecker() {
    loadDefaultRules();
}

ConstraintChecker::~ConstraintChecker() = default;

void ConstraintChecker::addRule(const ConstraintRule& rule) {
    std::lock_guard<std::mutex> lock(mutex_);
    rules_[rule.id] = rule;
}

void ConstraintChecker::removeRule(const std::string& ruleId) {
    std::lock_guard<std::mutex> lock(mutex_);
    rules_.erase(ruleId);
    customCheckers_.erase(ruleId);
}

void ConstraintChecker::setRuleEnabled(const std::string& ruleId, bool enabled) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = rules_.find(ruleId);
    if (it != rules_.end()) {
        it->second.enabled = enabled;
    }
}

std::vector<ConstraintRule> ConstraintChecker::getRules() const {
    std::lock_guard<std::mutex> lock(mutex_);
    std::vector<ConstraintRule> result;
    for (const auto& pair : rules_) {
        result.push_back(pair.second);
    }
    return result;
}

std::vector<ConstraintRule> ConstraintChecker::getRulesByType(ConstraintType type) const {
    std::lock_guard<std::mutex> lock(mutex_);
    std::vector<ConstraintRule> result;
    for (const auto& pair : rules_) {
        if (pair.second.type == type) {
            result.push_back(pair.second);
        }
    }
    return result;
}

ConstraintReport ConstraintChecker::check(const std::string& taskId,
                                           const std::string& skillId,
                                           const std::string& operation,
                                           const json& data) {
    ConstraintReport report;
    report.taskId = taskId;
    report.skillId = skillId;
    report.operation = operation;
    report.timestamp = nowMs();
    report.allPassed = true;

    std::lock_guard<std::mutex> lock(mutex_);

    // Get applicable rules
    std::vector<ConstraintRule> applicableRules;
    for (const auto& pair : rules_) {
        if (pair.second.enabled) {
            applicableRules.push_back(pair.second);
        }
    }

    // Check recursively
    checkJsonRecursive(data, "", report.results, applicableRules);

    // Determine overall result
    for (const auto& result : report.results) {
        if (!result.passed) {
            report.allPassed = false;
        }
    }

    return report;
}

ConstraintResult ConstraintChecker::checkField(const std::string& field,
                                                const json& value,
                                                const std::vector<ConstraintRule>& rules) {
    ConstraintResult result;
    result.field = field;
    result.passed = true;

    // Check each applicable rule
    for (const auto& rule : rules) {
        if (rule.fields.empty() || rule.fields.find(field) != rule.fields.end()) {
            // Check pattern match
            if (!rule.pattern.empty() && value.is_string()) {
                std::string strValue = value.get<std::string>();
                if (matchPattern(rule.pattern, strValue)) {
                    result.ruleId = rule.id;
                    result.ruleName = rule.name;
                    result.type = rule.type;
                    result.severity = rule.severity;
                    result.passed = false;
                    result.message = rule.description;
                    result.matchedValue = strValue;
                    break;
                }
            }

            // Check custom checker
            auto it = customCheckers_.find(rule.id);
            if (it != customCheckers_.end()) {
                if (!it->second(value)) {
                    result.ruleId = rule.id;
                    result.ruleName = rule.name;
                    result.type = rule.type;
                    result.severity = rule.severity;
                    result.passed = false;
                    result.message = rule.description;
                    break;
                }
            }
        }
    }

    return result;
}

void ConstraintChecker::setCustomChecker(const std::string& ruleId,
                                          std::function<bool(const json&)> checker) {
    std::lock_guard<std::mutex> lock(mutex_);
    customCheckers_[ruleId] = checker;
}

void ConstraintChecker::loadDefaultRules() {
    auto defaults = DefaultConstraints::getAllDefaults();
    for (const auto& rule : defaults) {
        addRule(rule);
    }
}

void ConstraintChecker::loadRules(const json& rulesJson) {
    std::lock_guard<std::mutex> lock(mutex_);

    if (rulesJson.is_array()) {
        for (const auto& ruleJson : rulesJson) {
            ConstraintRule rule = ConstraintRule::fromJson(ruleJson);
            rules_[rule.id] = rule;
        }
    }
}

json ConstraintChecker::exportRules() const {
    std::lock_guard<std::mutex> lock(mutex_);

    json j = json::array();
    for (const auto& pair : rules_) {
        j.push_back(pair.second.toJson());
    }
    return j;
}

void ConstraintChecker::clearRules() {
    std::lock_guard<std::mutex> lock(mutex_);
    rules_.clear();
    customCheckers_.clear();
}

void ConstraintChecker::checkJsonRecursive(const json& data,
                                            const std::string& prefix,
                                            std::vector<ConstraintResult>& results,
                                            const std::vector<ConstraintRule>& applicableRules) {
    if (data.is_object()) {
        for (auto it = data.begin(); it != data.end(); ++it) {
            std::string fieldPath = prefix.empty() ? it.key() : prefix + "." + it.key();
            checkJsonRecursive(it.value(), fieldPath, results, applicableRules);
        }
    } else if (data.is_array()) {
        for (size_t i = 0; i < data.size(); ++i) {
            std::string fieldPath = prefix + "[" + std::to_string(i) + "]";
            checkJsonRecursive(data[i], fieldPath, results, applicableRules);
        }
    } else {
        // Leaf value - check it
        ConstraintResult result = checkField(prefix, data, applicableRules);
        if (!result.passed || result.severity == ConstraintSeverity::Audit) {
            results.push_back(result);
        }
    }
}

bool ConstraintChecker::matchPattern(const std::string& pattern, const std::string& value) {
    try {
        std::regex re(pattern);
        return std::regex_search(value, re);
    } catch (...) {
        return false;
    }
}

// ==================== DefaultConstraints ====================

ConstraintRule DefaultConstraints::createPrivacyRule() {
    ConstraintRule rule;
    rule.id = "privacy-sensitive-fields";
    rule.name = "Sensitive Field Detection";
    rule.type = ConstraintType::Privacy;
    rule.severity = ConstraintSeverity::Block;
    rule.description = "Sensitive data field detected";
    rule.pattern = "";  // Pattern matching on field names instead
    rule.fields = {
        "password", "passwd", "pwd", "secret", "token", "api_key",
        "credit_card", "ssn", "social_security", "phone", "email",
        "address", "birth_date", "name"
    };
    rule.enabled = true;
    rule.createdAt = nowMs();
    return rule;
}

ConstraintRule DefaultConstraints::createFinancialRule() {
    ConstraintRule rule;
    rule.id = "financial-validation";
    rule.name = "Financial Data Validation";
    rule.type = ConstraintType::Financial;
    rule.severity = ConstraintSeverity::Audit;
    rule.description = "Financial transaction detected";
    rule.pattern = "\\d{4}-\\d{4}-\\d{4}-\\d{4}";  // Credit card pattern
    rule.fields = {"credit_card", "card_number", "account"};
    rule.enabled = true;
    rule.createdAt = nowMs();
    return rule;
}

ConstraintRule DefaultConstraints::createSecurityRule() {
    ConstraintRule rule;
    rule.id = "security-command-check";
    rule.name = "Security Command Check";
    rule.type = ConstraintType::Security;
    rule.severity = ConstraintSeverity::Block;
    rule.description = "Potential command injection detected";
    rule.pattern = "(exec|eval|system|shell|cmd|command).*";
    rule.fields = {"command", "cmd", "script", "code"};
    rule.enabled = true;
    rule.createdAt = nowMs();
    return rule;
}

std::vector<ConstraintRule> DefaultConstraints::getAllDefaults() {
    std::vector<ConstraintRule> rules;
    rules.push_back(createPrivacyRule());
    rules.push_back(createFinancialRule());
    rules.push_back(createSecurityRule());

    // Auth rule
    ConstraintRule authRule;
    authRule.id = "auth-required";
    authRule.name = "Authentication Required";
    authRule.type = ConstraintType::Auth;
    authRule.severity = ConstraintSeverity::Block;
    authRule.description = "Operation requires authentication";
    authRule.enabled = true;
    authRule.createdAt = nowMs();
    rules.push_back(authRule);

    // Location rule
    ConstraintRule locRule;
    locRule.id = "location-restriction";
    locRule.name = "Location Restriction";
    locRule.type = ConstraintType::Location;
    locRule.severity = ConstraintSeverity::Warning;
    locRule.description = "Location-based operation restricted";
    locRule.enabled = true;
    locRule.createdAt = nowMs();
    rules.push_back(locRule);

    // Time rule
    ConstraintRule timeRule;
    timeRule.id = "time-restriction";
    timeRule.name = "Time Restriction";
    timeRule.type = ConstraintType::Time;
    timeRule.severity = ConstraintSeverity::Warning;
    timeRule.description = "Operation restricted by time";
    timeRule.enabled = true;
    timeRule.createdAt = nowMs();
    rules.push_back(timeRule);

    return rules;
}

} // namespace ofa

File: src/sdk/go/constraint.go
package ofa

import (
	"encoding/json"
	"fmt"
	"regexp"
	"sync"
	"time"
)

// ConstraintType defines constraint type
type ConstraintType int

const (
	ConstraintPrivacy ConstraintType = iota
	ConstraintFinancial
	ConstraintSecurity
	ConstraintAuth
	ConstraintLocation
	ConstraintTime
	ConstraintCustom
)

// ConstraintSeverity defines constraint severity
type ConstraintSeverity int

const (
	SeverityWarning ConstraintSeverity = iota
	SeverityBlock
	SeverityAudit
)

// ConstraintRule defines a constraint rule
type ConstraintRule struct {
	ID          string            `json:"id"`
	Name        string            `json:"name"`
	Type        ConstraintType    `json:"type"`
	Severity    ConstraintSeverity `json:"severity"`
	Description string            `json:"description"`
	Pattern     string            `json:"pattern"`
	Fields      []string          `json:"fields"`
	Enabled     bool              `json:"enabled"`
	CreatedAt   int64             `json:"createdAt"`
}

// ConstraintResult represents a constraint check result
type ConstraintResult struct {
	RuleID       string            `json:"ruleId"`
	RuleName     string            `json:"ruleName"`
	Type         ConstraintType    `json:"type"`
	Severity     ConstraintSeverity `json:"severity"`
	Passed       bool              `json:"passed"`
	Message      string            `json:"message"`
	Field        string            `json:"field"`
	MatchedValue string            `json:"matchedValue"`
}

// ConstraintReport represents a full constraint report
type ConstraintReport struct {
	TaskID    string             `json:"taskId"`
	SkillID   string             `json:"skillId"`
	Operation string             `json:"operation"`
	AllPassed bool               `json:"allPassed"`
	Results   []*ConstraintResult `json:"results"`
	Timestamp int64              `json:"timestamp"`
}

// HasBlockers checks if report has blocking results
func (r *ConstraintReport) HasBlockers() bool {
	for _, result := range r.Results {
		if !result.Passed && result.Severity == SeverityBlock {
			return true
		}
	}
	return false
}

// GetBlockerMessages returns blocking messages
func (r *ConstraintReport) GetBlockerMessages() []string {
	var messages []string
	for _, result := range r.Results {
		if !result.Passed && result.Severity == SeverityBlock {
			messages = append(messages, result.Message)
		}
	}
	return messages
}

// CustomChecker is a custom constraint checker function
type CustomChecker func(json.RawMessage) bool

// ConstraintChecker manages constraint checking
type ConstraintChecker struct {
	rules         map[string]*ConstraintRule
	customCheckers map[string]CustomChecker
	mu            sync.Mutex
}

// NewConstraintChecker creates a new constraint checker
func NewConstraintChecker() *ConstraintChecker {
	c := &ConstraintChecker{
		rules:         make(map[string]*ConstraintRule),
		customCheckers: make(map[string]CustomChecker),
	}
	c.loadDefaultRules()
	return c
}

// AddRule adds a constraint rule
func (c *ConstraintChecker) AddRule(rule *ConstraintRule) {
	c.mu.Lock()
	rule.CreatedAt = time.Now().UnixMilli()
	c.rules[rule.ID] = rule
	c.mu.Unlock()
}

// RemoveRule removes a rule
func (c *ConstraintChecker) RemoveRule(ruleID string) {
	c.mu.Lock()
	delete(c.rules, ruleID)
	delete(c.customCheckers, ruleID)
	c.mu.Unlock()
}

// SetRuleEnabled enables or disables a rule
func (c *ConstraintChecker) SetRuleEnabled(ruleID string, enabled bool) {
	c.mu.Lock()
	if rule, ok := c.rules[ruleID]; ok {
		rule.Enabled = enabled
	}
	c.mu.Unlock()
}

// GetRules returns all rules
func (c *ConstraintChecker) GetRules() []*ConstraintRule {
	c.mu.Lock()
	defer c.mu.Unlock()

	var rules []*ConstraintRule
	for _, rule := range c.rules {
		rules = append(rules, rule)
	}
	return rules
}

// GetRulesByType returns rules by type
func (c *ConstraintChecker) GetRulesByType(typ ConstraintType) []*ConstraintRule {
	c.mu.Lock()
	defer c.mu.Unlock()

	var rules []*ConstraintRule
	for _, rule := range c.rules {
		if rule.Type == typ {
			rules = append(rules, rule)
		}
	}
	return rules
}

// Check checks data against all rules
func (c *ConstraintChecker) Check(taskID, skillID, operation string, data json.RawMessage) *ConstraintReport {
	report := &ConstraintReport{
		TaskID:    taskID,
		SkillID:   skillID,
		Operation: operation,
		Timestamp: time.Now().UnixMilli(),
		AllPassed: true,
	}

	c.mu.Lock()
	applicableRules := c.getApplicableRules()
	c.mu.Unlock()

	c.checkJSONRecursive(data, "", applicableRules, report)

	for _, result := range report.Results {
		if !result.Passed {
			report.AllPassed = false
		}
	}

	return report
}

// SetCustomChecker sets a custom checker for a rule
func (c *ConstraintChecker) SetCustomChecker(ruleID string, checker CustomChecker) {
	c.mu.Lock()
	c.customCheckers[ruleID] = checker
	c.mu.Unlock()
}

// LoadRules loads rules from JSON
func (c *ConstraintChecker) LoadRules(data json.RawMessage) error {
	var rules []*ConstraintRule
	if err := json.Unmarshal(data, &rules); err != nil {
		return err
	}

	c.mu.Lock()
	for _, rule := range rules {
		c.rules[rule.ID] = rule
	}
	c.mu.Unlock()

	return nil
}

// ExportRules exports rules to JSON
func (c *ConstraintChecker) ExportRules() json.RawMessage {
	c.mu.Lock()
	defer c.mu.Unlock()

	rules := make([]*ConstraintRule, 0, len(c.rules))
	for _, rule := range c.rules {
		rules = append(rules, rule)
	}

	data, _ := json.Marshal(rules)
	return data
}

// ClearRules clears all rules
func (c *ConstraintChecker) ClearRules() {
	c.mu.Lock()
	c.rules = make(map[string]*ConstraintRule)
	c.customCheckers = make(map[string]CustomChecker)
	c.mu.Unlock()
}

func (c *ConstraintChecker) getApplicableRules() []*ConstraintRule {
	var rules []*ConstraintRule
	for _, rule := range c.rules {
		if rule.Enabled {
			rules = append(rules, rule)
		}
	}
	return rules
}

func (c *ConstraintChecker) checkJSONRecursive(data json.RawMessage, prefix string, rules []*ConstraintRule, report *ConstraintReport) {
	var v interface{}
	if err := json.Unmarshal(data, &v); err != nil {
		return
	}

	switch val := v.(type) {
	case map[string]interface{}:
		for key, value := range val {
			fieldPath := prefix + "." + key
			if prefix == "" {
				fieldPath = key
			}
			subData, _ := json.Marshal(value)
			c.checkJSONRecursive(subData, fieldPath, rules, report)
		}
	case []interface{}:
		for i, value := range val {
			fieldPath := fmt.Sprintf("%s[%d]", prefix, i)
			subData, _ := json.Marshal(value)
			c.checkJSONRecursive(subData, fieldPath, rules, report)
		}
	default:
		// Leaf value - check it
		result := c.checkField(prefix, data, rules)
		if !result.Passed || result.Severity == SeverityAudit {
			report.Results = append(report.Results, result)
		}
	}
}

func (c *ConstraintChecker) checkField(field string, value json.RawMessage, rules []*ConstraintRule) *ConstraintResult {
	result := &ConstraintResult{
		Field:  field,
		Passed: true,
	}

	c.mu.Lock()
	defer c.mu.Unlock()

	for _, rule := range rules {
		// Check if rule applies to this field
		if len(rule.Fields) > 0 {
			fieldMatch := false
			for _, f := range rule.Fields {
				if f == field || regexpMatch(f, field) {
					fieldMatch = true
					break
				}
			}
			if !fieldMatch {
				continue
			}
		}

		// Check pattern
		if rule.Pattern != "" {
			var strValue string
			var v interface{}
			if json.Unmarshal(value, &v) == nil {
				if s, ok := v.(string); ok {
					strValue = s
				}
			}

			if strValue != "" && regexpMatch(rule.Pattern, strValue) {
				result.RuleID = rule.ID
				result.RuleName = rule.Name
				result.Type = rule.Type
				result.Severity = rule.Severity
				result.Passed = false
				result.Message = rule.Description
				result.MatchedValue = strValue
				return result
			}
		}

		// Check custom checker
		if checker, ok := c.customCheckers[rule.ID]; ok {
			if !checker(value) {
				result.RuleID = rule.ID
				result.RuleName = rule.Name
				result.Type = rule.Type
				result.Severity = rule.Severity
				result.Passed = false
				result.Message = rule.Description
				return result
			}
		}
	}

	return result
}

func regexpMatch(pattern, value string) bool {
	re, err := regexp.Compile(pattern)
	if err != nil {
		return false
	}
	return re.MatchString(value)
}

func (c *ConstraintChecker) loadDefaultRules() {
	// Privacy rule
	c.AddRule(&ConstraintRule{
		ID:          "privacy-sensitive-fields",
		Name:        "Sensitive Field Detection",
		Type:        ConstraintPrivacy,
		Severity:    SeverityBlock,
		Description: "Sensitive data field detected",
		Fields: []string{
			"password", "passwd", "pwd", "secret", "token", "api_key",
			"credit_card", "ssn", "social_security", "phone", "email",
			"address", "birth_date", "name",
		},
		Enabled: true,
	})

	// Financial rule
	c.AddRule(&ConstraintRule{
		ID:          "financial-validation",
		Name:        "Financial Data Validation",
		Type:        ConstraintFinancial,
		Severity:    SeverityAudit,
		Description: "Financial transaction detected",
		Pattern:     "\\d{4}-\\d{4}-\\d{4}-\\d{4}",
		Fields:      []string{"credit_card", "card_number", "account"},
		Enabled:     true,
	})

	// Security rule
	c.AddRule(&ConstraintRule{
		ID:          "security-command-check",
		Name:        "Security Command Check",
		Type:        ConstraintSecurity,
		Severity:    SeverityBlock,
		Description: "Potential command injection detected",
		Pattern:     "(exec|eval|system|shell|cmd|command).*",
		Fields:      []string{"command", "cmd", "script", "code"},
		Enabled:     true,
	})

	// Auth rule
	c.AddRule(&ConstraintRule{
		ID:          "auth-required",
		Name:        "Authentication Required",
		Type:        ConstraintAuth,
		Severity:    SeverityBlock,
		Description: "Operation requires authentication",
		Enabled:     true,
	})

	// Location rule
	c.AddRule(&ConstraintRule{
		ID:          "location-restriction",
		Name:        "Location Restriction",
		Type:        ConstraintLocation,
		Severity:    SeverityWarning,
		Description: "Location-based operation restricted",
		Enabled:     true,
	})

	// Time rule
	c.AddRule(&ConstraintRule{
		ID:          "time-restriction",
		Name:        "Time Restriction",
		Type:        ConstraintTime,
		Severity:    SeverityWarning,
		Description: "Operation restricted by time",
		Enabled:     true,
	})
}